src/mcts/search.cc

```diff
@@ -818,48 +847,62 @@ int SearchWorker::PrefetchIntoCache(Node* node, int budget) {
-  // This function is so small as to be silly, but its parent function is
-  // conceptually cleaner for it.
-  if (computation_->GetBatchSize() != 0) computation_->ComputeBlocking();
+  // Do not do NN eval if there's nothing to compute, or if it's an out-of-order
+  // eval, don't compute.
+  if (!out_of_order_eval_ && computation_->GetBatchSize() != 0) {
+    computation_->ComputeBlocking();
+  }
 }
 
 // 5. Retrieve NN computations (and terminal values) into nodes.
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 void SearchWorker::FetchMinibatchResults() {
+  // If it is an out-of-order eval, only eval last item in a batch.
+  if (out_of_order_eval_) {
+    FetchSingleNodeResult(&nodes_to_process_.back(),
+                          computation_->GetBatchSize() - 1);
+    return;
+  }
+
   // Populate NN/cached results, or terminal results, into nodes.
   int idx_in_computation = 0;
   for (auto& node_to_process : nodes_to_process_) {
-    Node* node = node_to_process.node;
-    if (!node_to_process.nn_queried) {
-      // Terminal nodes don't involve the neural NetworkComputation, nor do
-      // they require any further processing after value retrieval.
-      node_to_process.v = node->GetQ();
-      continue;
-    }
-    // For NN results, we need to populate policy as well as value.
-    // First the value...
-    node_to_process.v = -computation_->GetQVal(idx_in_computation);
-    // ...and secondly, the policy data.
-    float total = 0.0;
-    for (auto edge : node->Edges()) {
-      float p = computation_->GetPVal(idx_in_computation,
-                                      edge.GetMove().as_nn_index());
-      if (search_->kPolicySoftmaxTemp != 1.0f) {
-        p = pow(p, 1 / search_->kPolicySoftmaxTemp);
-      }
-      total += p;
-      edge.edge()->SetP(p);
-    }
-    // Normalize P values to add up to 1.0.
-    if (total > 0.0f) {
-      float scale = 1.0f / total;
-      for (auto edge : node->Edges()) edge.edge()->SetP(edge.GetP() * scale);
-    }
-    // Add Dirichlet noise if enabled and at root.
-    if (search_->kNoise && node == search_->root_node_) {
-      ApplyDirichletNoise(node, 0.25, 0.3);
+    FetchSingleNodeResult(&node_to_process, idx_in_computation);
+    if (node_to_process.nn_queried) ++idx_in_computation;
+  }
+}
+
+void SearchWorker::FetchSingleNodeResult(NodeToProcess* node_to_process,
+                                         int idx_in_computation) {
+  Node* node = node_to_process->node;
+  if (!node_to_process->nn_queried) {
+    // Terminal nodes don't involve the neural NetworkComputation, nor do
+    // they require any further processing after value retrieval.
+    node_to_process->v = node->GetQ();
+    return;
+  }
+  // For NN results, we need to populate policy as well as value.
+  // First the value...
+  node_to_process->v = -computation_->GetQVal(idx_in_computation);
+  // ...and secondly, the policy data.
+  float total = 0.0;
+  for (auto edge : node->Edges()) {
```

Comment:
It occurs to me now, for cached NN results, we recompute the policy normalization, and if non-unit, we recompute the softmax temperature each time as well. Why isn't this done directly in the CachingComputation and stored as part of the cache, so we don't re-do these policy shenanigans for every cache hit?

(This comment is totally ancillary to the rest of the review)

Reply:
Caching computation doesn't always know the list of legal moves. In prefetch-into-cache, it stores values for all pseudo-legal moves instead.
It's a good optimization to keep in mind, especially when we get rid of history, but it's not that easy to implement nicely.